### PR TITLE
New module - meraki_mx_l7_firewall

### LINKED ISSUE #55450: New module - meraki_mx_l7_firewall
##### SUMMARY
Meraki recently added support for listing and manipulating layer 7 firewall rules via API. A new module should be created to support these endpoints.

https://dashboard.meraki.com/api_docs#mx-l7-firewall

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
meraki

### PR DESCRIPTION
**Pull request by kbreit**
##### SUMMARY
This new module allows modification and querying of MX appliance layer 7 firewall rules.

@brianberzins

Fixes #55450

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
meraki_mx_l7_firewall

##### ADDITIONAL INFORMATION

```paste below
  - name: Create multiple firewall rules
    meraki_mx_l7_firewall:
      auth_key: '{{ auth_key }}'
      org_name: '{{test_org_name}}'
      net_name: TestNetAppliance
      state: present
      rules:
        - type: whitelisted_countries
          countries:
            - US
            - FR
        - type: blacklisted_countries
          countries:
            - CN
        - policy: deny
          type: port
          port: 8080
    delegate_to: localhost

TASK [meraki_mx_l7_firewall : debug] *******************************************
ok: [localhost] => {
        "data": {
            "rules": [
                {
                    "policy": "deny",
                    "type": "whitelistedCountries",
                    "value": [
                        "US",
                        "FR"
                    ]
                },
                {
                    "policy": "deny",
                    "type": "blacklistedCountries",
                    "value": [
                        "CN"
                    ]
                },
                {
                    "policy": "deny",
                    "type": "port",
                    "value": "8080"
                }
            ]
        },
```

### CODE REVIEW

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py [hunk 1]
```diff
@@ -0,0 +1,512 @@
...
+    net_name:
+        description:
+        - Name of network which MX firewall is in.
+    net_id:
+        description:
+        - ID of network which MX firewall is in.
+    rules:
+        description:
+        - List of layer 7 firewall rules.
+        suboptions:
```
Comment (Qalthos) [inline lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py:45]:
```suggestion
        type: list
        suboptions:
```

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+                        description: Descriptive name of application category.
+                        returned: success
+                        type: string
+                        sample: layer7/category/1
+'''
+
+import copy
+import os
+from ansible.module_utils.basic import AnsibleModule, json, env_fallback
+from ansible.module_utils.urls import fetch_url
```
Comment (Qalthos) [inline lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py]: Are you using this somewhere?

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+from ansible.module_utils._text import to_native
```
Comment (Qalthos) [inline lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py]: Or this?

**Review by Qalthos — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+notes:
+- Module assumes a complete list of firewall rules are passed as a parameter.
+- If there is interest in this module allowing manipulation of a single firewall rule, please submit an issue against this module.
+options:
+    state:
+        description:
+        - Create or modify an organization.
+        choices: ['present', 'query']
+        default: present
+    org_name:
```
Comment (Qalthos) [inline lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py]: `org_name` and `org_id` are already defined in the meraki doc_fragment. I can understand overriding it as the description differs from there (though I do not know immediately which would take precedence), but there is also no `clone` parameter documented

**Review by Qalthos — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py (hunk 1 above)
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py:45]: Updated alongside other changes.

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py (hunk 2 above)
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py]: Removed.

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py (hunk 3 above)
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py]: Removed.

File: lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py (hunk 4 above)
Comment (kbreit) [reply lib/ansible/modules/network/meraki/meraki_mx_l7_firewall.py]: I just ran a test. `make webdocs` will prefer the documentation in the module, if provided. I did remove `org_name` and `org_id` to simplify though. `net_*` is still there.

**Comment by kbreit** [comment]
@Qalthos Please take a look at this PR as it should include all your feedback.

**Review by Qalthos — APPROVED** [review]

**Review by pabelanger — APPROVED** [review]

**Pull request merged**